### PR TITLE
`General`: Merge scores and submission pages into one page for exercise instructors

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The submission and scores tables of each exercise nearly showed the same information but created many duplications.

### Description

This PR merges these to tables in the following way:
- The submission table and the buttons redirecting to it are removed. If any place needs a redirect to the scores table and does not already have it, let me know.
- The scores table always shows all participations with a submission and in case of a submission without result displays "no result" instead.
- The filter now have more options and are moved into one drop down instead of multiple options next to each other
- The buttons redirecting to the manual assessment (and code editor) are now consistently shown in each exercise and configuration as long as manual assessments or editing code is possible.
- Admins can no longer go to the user editing page from the scores table.
- The assessment overview does not have a link to the own assessed submissions, but all assessments are still visible in the assessment overview

### Steps for Testing

Prerequisites:
- 1 Instructor
- 3 Students
- 1 Exercise of each type

1. Log in to Artemis
2. Participate with each student in the exercises, on only starts the exercise, one only submits something, one also gets an assessment
3. Check the scores page and all related views like the exercise overview or the assessment view

#### Exam Mode Testing

Same as normal mode, but with the exercises in an exam

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| exercise-scores.component.ts | 75% | 90% |
| exercise.utils.ts | 90% | 88% |

#### Programming exercises (automatic but with complaints):
Old scores table:
Bildschirm­foto 2023-03-14 um 21 02 38
Old submissions table:
Bildschirm­foto 2023-03-14 um 21 02 45
New merged table:
Bildschirm­foto 2023-03-15 um 11 03 16
#### Modeling/Text/Fileupload exercises:
Old scores table:
Bildschirm­foto 2023-03-15 um 09 28 18
Old submissions table:
Bildschirm­foto 2023-03-15 um 09 28 32
New merged table:
Bildschirm­foto 2023-03-15 um 09 24 35
#### Quiz exercises:
Did not have a submission table.
New merged table:
Bildschirm­foto 2023-03-15 um 11 15 22
#### Filter and export:
Bildschirm­foto 2023-03-15 um 11 03 23
Bildschirm­foto 2023-03-15 um 11 06 25

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html [hunk 1]
```diff
@@ -72,20 +68,18 @@
...
-                            <a [routerLink]="submissionsLink" [queryParams]="{ tutorId: tutor?.id }">{{ numberOfTutorAssessments }} ({{ tutorAssessmentPercentage }}%)</a>
+                            {{ numberOfTutorAssessments }} ({{ tutorAssessmentPercentage }}%)
                         </div>
                         <div *ngIf="!exercise.isAtLeastInstructor">
                             <span class="text-success">{{ numberOfSubmissions.inTime }}</span>
                             <span *ngIf="numberOfSubmissions.late > 0"> - </span>
                             <span *ngIf="numberOfSubmissions.late > 0" class="text-danger">{{ numberOfSubmissions.late }}</span>
                             |
-                            <a [routerLink]="submissionsLink" [queryParams]="{ tutorId: tutor?.id }">{{ numberOfTutorAssessments }} ({{ tutorAssessmentPercentage }}%)</a>
+                            {{ numberOfTutorAssessments }} ({{ tutorAssessmentPercentage }}%)
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html:82]: This is now duplicated (the same code for instructors and tutors), can you merge that?

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html [hunk 2]
```diff
@@ -155,75 +146,80 @@ <h4 *ngIf="this.rangeFilter" class="d-flex align-items-center">
...
-                            <fa-icon [icon]="controls.iconForSortPropField('participation.student.login')"></fa-icon>
+                            <fa-icon [icon]="controls.iconForSortPropField('student.login')"></fa-icon>
                         </span>
                     </ng-template>
                     <ng-template ngx-datatable-cell-template let-value="value">
                         <a *ngIf="exercise.teamMode; else defaultMode" [routerLink]="['/course-management', exercise.course?.id, 'exercises', exercise.id, 'teams', value.team.id]">
                             {{ value.team.shortName }}
                         </a>
                         <ng-template #defaultMode>
-                            <a *ngIf="isAdmin; else showLogin" [routerLink]="['/admin', 'user-management', value.student.login]">
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html:174]: Why did you remove the user management links here?

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts [hunk 3]
```diff
@@ -249,21 +215,21 @@ export class ExerciseScoresComponent implements OnInit, OnDestroy {
     }
 
     /**
-     * Returns the link to the repository of a result
-     * @param result Result for which to get the link for
+     * Returns the link to the repository of a participation
+     * @param participation Üarticipation for which to get the link for
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts:219]:
```suggestion
     * @param participation Participation for which to get the link for
```

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts [hunk 4]
```diff
@@ -18,25 +19,24 @@ import { ExerciseService } from 'app/exercises/shared/exercise/exercise.service'
...
 import { ExerciseCacheService } from 'app/exercises/shared/exercise/exercise-cache.service';
 
 /**
  * Filter properties for a result
  */
-enum FilterProp {
+export enum FilterProp {
     ALL = 'all',
-    SUCCESSFUL = 'successful',
-    UNSUCCESSFUL = 'unsuccessful',
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts:36]: Can we keep the successful / unsuccessful filters? I use them regularly.

File: src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts [hunk 5]
```diff
@@ -195,11 +195,7 @@ export const areManualResultsAllowed = (exercise: Exercise) => {
     // Only allow new results if manual reviews are activated and the due date/after due date has passed
     const progEx = exercise as ProgrammingExercise;
     const relevantDueDate = progEx.buildAndTestStudentSubmissionsAfterDueDate ?? progEx.dueDate;
-    return (
-        (!!progEx.isAtLeastTutor || !!progEx.isAtLeastEditor || !!progEx.isAtLeastInstructor) &&
-        progEx.assessmentType === AssessmentType.SEMI_AUTOMATIC &&
-        (!relevantDueDate || dayjs(relevantDueDate).isBefore(dayjs()))
-    );
+    return !!exercise.isAtLeastTutor && progEx.assessmentType === AssessmentType.SEMI_AUTOMATIC && (!relevantDueDate || dayjs(relevantDueDate).isBefore(dayjs()));
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts:198]: Is this logic correct in case the due date is null? Should I be able to create a new manual result even though the student can still push changes?

Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts:198]: You're right

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts (hunk 4 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts:36]: Readded, now as dropdown for all filters

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html [hunk 6]
```diff
@@ -238,205 +262,86 @@ <h4 *ngIf="this.rangeFilter" class="d-flex align-items-center">
                     </ng-template>
                 </ngx-datatable-column>
 
-                <ngx-datatable-column prop="durationInMinutes" [resizeable]="false" [sortable]="true" [minWidth]="100" [width]="100">
+                <ngx-datatable-column prop="" [sortable]="true" [minWidth]="80" [width]="90">
```
Comment (b-fein) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html:265]: Specifying `width` probably immediately overwrites the `minWidth` property.

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html [hunk 7]
```diff
@@ -238,205 +262,86 @@ <h4 *ngIf="this.rangeFilter" class="d-flex align-items-center">
...
+                                    [routerLink]="['/courses', course.id, 'programming-exercises', exercise.id, 'code-editor', value.id]"
+                                    class="btn btn-info btn-sm me-1"
+                                    [jhiFeatureToggleLink]="FeatureToggle.ProgrammingExercises"
+                                >
+                                    <fa-icon class="me-1" [icon]="faFolderOpen" [fixedWidth]="true"></fa-icon>
+                                    Online editor
+                                </a>
+                            </div>
+                            <div>
+                                <a
```
Comment (b-fein) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html:297]: The submissions table had a green button ‘Assess Submissions’ in case no manual submission existed, and changed that to a blue one ‘View Assessment’ in case one did. It also showed ’Continue Assessment’/’Cancel Assessment’ for locked but unsubmitted ones.

Can you port this functionality over to the new table (maybe extract it into its own small component in case it wasn’t already)? That feature was quite useful for us when not using double-blind reviews. Ideally it would also show the reviewer.

**Review by b-fein — CHANGES_REQUESTED** [review]
Thanks for simplifying the tables and removing unnecessary clutter for both developers and users. :+1: 

The ‘View Assessment’ button on the submissions table was a lot more powerful compared to the one that was kept on the scores table. I regularly used its features and would suggest moving it over to the new combined table.

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html (hunk 6 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html:265]: This is done in many components. Afaik the width only specifies the preferred width, but in case of small screens, the width can be shrunken down to the min width

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html (hunk 7 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html:297]: There is still the option to get to the assessment of each result as well as start one. I don't really see any benefit of distinguishing between the two cases, do you?

How often did you use the cancel/continue functionality?
- Continuing is easier from the assessment overview in my opinion or did you specifically search for a student where you wanted to continue the assessment?
- Canceling is readded :+1:

Comment (b-fein) [reply src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.html:297]: > There is still the option to get to the assessment of each result as well as start one. I don't really see any benefit of distinguishing between the two cases, do you?

With the different colour it was immediately obvious if the student already has gotten a manual assessment, or is missing one. Before it was kind-of redundant since the result type and the assessor were shown in additional columns as well, but now. Now this information is not presented any more, unless you open the feedback view for the result to check if manual feedback exist.
The cases with an unfinished assessment described below would become immediately obvious when the different states were visually different.

> Continuing is easier from the assessment overview in my opinion or did you specifically search for a student where you wanted to continue the assessment?

Yes, we do not resolve plagiarism cases via Artemis but require students to come in person to the office to discuss those matters. In those cases an instructors locks those two submissions by starting the assessment, but not finishing it, so that it does not show up to other assessors and the assessments can continue for other submissions. Only after the 
discussion exactly those two assessments need to be continued and submitted.

We had at least once were where some tutor started an assessment but forgot to submit it. Then they were not reachable (e.g. holidays) and someone else had to look over and finish it. In that case continuing from the assessment dashboard would not have been possible, I think, as only the original assessor can do that.


> Canceling is readded :+1:

Thanks!

**Review by nikirovast — DISMISSED** [review]
Tested on ts1 for usual exercises as well as for the exam. Seems to work!

**Review by YeetTheFirst21 — APPROVED** [review]
Tested on TS1 with all exercises, seems to be working without issues

**Review by Strohgelaender — APPROVED** [review]
code changes look good to me

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html [hunk 8]
```diff
@@ -71,22 +67,11 @@
...
-                            <a [routerLink]="submissionsLink" [queryParams]="{ tutorId: tutor?.id }">{{ numberOfTutorAssessments }} ({{ tutorAssessmentPercentage }}%)</a>
-                        </div>
-                        <div *ngIf="!exercise.isAtLeastInstructor">
-                            <span class="text-success">{{ numberOfSubmissions.inTime }}</span>
-                            <span *ngIf="numberOfSubmissions.late > 0"> - </span>
-                            <span *ngIf="numberOfSubmissions.late > 0" class="text-danger">{{ numberOfSubmissions.late }}</span>
-                            |
-                            <a [routerLink]="submissionsLink" [queryParams]="{ tutorId: tutor?.id }">{{ numberOfTutorAssessments }} ({{ tutorAssessmentPercentage }}%)</a>
-                        </div>
+                        <span class="text-success">{{ numberOfSubmissions.inTime }}</span>
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html:70]: probably should be a `<p>` for this entire text

File: src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts [hunk 9]
```diff
@@ -178,23 +178,19 @@ export const isStartPracticeAvailable = (exercise: Exercise, participation?: Stu
...
+ * If it is a programming exercise, the due date also has to be in the past.
  *
  * @param exercise
  */
-export const areManualResultsAllowed = (exercise: Exercise) => {
-    if (exercise.type !== ExerciseType.PROGRAMMING) {
+export const areManualResultsAllowed = (exercise: Exercise): boolean => {
+    if (exercise.type === ExerciseType.QUIZ) {
         return false;
+    } else if (exercise.type !== ExerciseType.PROGRAMMING) {
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts:188]: redundant else after return; The following else as well

**Review by simon-hng — APPROVED** [review]
Left some minor suggestions. Code looks good otherwise

File: src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts (hunk 9 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/exercise/exercise.utils.ts:188]: It's redundant but makes it even clearer that only one option is possible, makes sure that changing the return to something else will still only go into one case and also produces shorter code

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html (hunk 8 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html:70]: I did not change the styling here and would like to keep it that way

**Review by tobias-lippert — DISMISSED** [review]
Code LGTM

**Review by Strohgelaender — APPROVED** [review]
code looks good.

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts1. Worked as expected.

**Review by danielbode — APPROVED** [review]
Tested the exam mode during a testing session on ts1. Works as expected.

**Review by DominikRemo — APPROVED** [review]
Tested in testing session. Tested on ts1. Worked as intended.

**Review by Strohgelaender — APPROVED** [review]

**Review by tobias-lippert — DISMISSED** [review]
reapprove

**Review by danielbode — APPROVED** [review]
Reapproved

**Review by DominikRemo — DISMISSED** [review]
Reapproved

File: src/main/webapp/app/assessment/assessment-filters/assessment-filters.component.ts [hunk 10]
```diff
@@ -1,65 +0,0 @@
...
-        this.filterProp = filterProp;
-        this.updateFilteredSubmissions();
-    }
-
-    private updateFilteredSubmissions() {
-        this.filterChange.emit(this.submissions.filter(this.filterSubmissionByProp));
-    }
-
-    private filterSubmissionByProp = (submission: Submission) => {
-        switch (this.filterProp) {
```
Comment (simon-hng) [inline src/main/webapp/app/assessment/assessment-filters/assessment-filters.component.ts:53]: if-else would also work here and possibly be simpler

File: src/main/webapp/app/course/dashboards/assessment-dashboard/assessment-dashboard.component.html [hunk 11]
```diff
@@ -206,9 +206,6 @@ <h4 jhiTranslate="artemisApp.assessmentDashboard.exerciseTable.title"></h4>
                                         [togglingSecondCorrectionButton]="isTogglingSecondCorrection.get(exercise.id!)!"
                                     ></jhi-second-correction-enable-button>
                                 </ng-container>
-                                <a class="btn btn-primary btn-sm mb-1" *ngIf="course?.isAtLeastInstructor" [routerLink]="getSubmissionsLinkForExercise(exercise)">
```
Comment (simon-hng) [inline src/main/webapp/app/course/dashboards/assessment-dashboard/assessment-dashboard.component.html:209]: If it has the css class btn, looks and works like a button; it probably should be a `<button>`

File: src/main/webapp/app/course/manage/overview/course-management-exercise-row.component.html [hunk 12]
```diff
@@ -145,13 +145,6 @@ <h4 class="fw-medium" *ngFor="let category of details.categories">
                     [ngbTooltip]="'entity.action.scores' | artemisTranslate"
                     ><fa-icon [icon]="faTable"></fa-icon
                 ></a>
-                <a
```
Comment (simon-hng) [inline src/main/webapp/app/course/manage/overview/course-management-exercise-row.component.html:148]: <button>? Links above and below probably as well

File: src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment-dashboard.component.html [hunk 13]
```diff
@@ -1,204 +0,0 @@
...
-                                            >
-                                                <fa-icon [fixedWidth]="true" [icon]="faBan"></fa-icon>&nbsp;{{
-                                                    'artemisApp.assessment.dashboard.actions.cancel' | artemisTranslate
-                                                }}
-                                            </button>
-                                        </div>
-                                    </span>
-                                </span>
-                                <!-- button for submission without result for this correction round-->
-                                <a
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment-dashboard.component.html:183]: button?

File: src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise.component.html [hunk 14]
```diff
@@ -95,14 +95,6 @@ <h4 class="d-flex mb-3 flex-wrap justify-content-end">
                                 <fa-icon [icon]="farListAlt"></fa-icon>
                                 <span class="d-none d-md-inline" jhiTranslate="artemisApp.exercise.participations">Participations</span>
                             </a>
-                            <a
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/file-upload/manage/file-upload-exercise.component.html:98]: button

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html (hunk 8 above)
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.html:70]: How about using paragraph? `<p>`

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.scss [hunk 15]
```diff
@@ -1,16 +0,0 @@
-.exercise-scores-filter-toolbar {
-    /*
-      Window width required for same-line display of filters and buttons by site language:
-        - English: 1400px
-        - German: 1870px
-     */
-    @media (min-width: 1400px) {
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.scss:7]: I don't think this is a good solution to hard-code it like this. What if a user zooms in or has a another font size set in his browser?

**Review by simon-hng — APPROVED** [review]
Code looks good. Left some minor suggestions for a possible follow up to improve usability and accessibility
